File: ai-engine/analyzers/pattern_detector.py
import pandas as pd
import numpy as np

def detect_candlestick_patterns(df):
    patterns = []
    
    # Get last few candles
    if len(df) < 3:
        return patterns
    
    last = df.iloc[-1]
    prev = df.iloc[-2]
    prev2 = df.iloc[-3] if len(df) >= 3 else None
    
    body = abs(last['close'] - last['open'])
    upper_shadow = last['high'] - max(last['open'], last['close'])
    lower_shadow = min(last['open'], last['close']) - last['low']
    
    # Hammer
    if lower_shadow > body * 2 and upper_shadow < body * 0.3 and last['close'] > last['open']:
        patterns.append('Hammer')
    
    # Shooting Star
    if upper_shadow > body * 2 and lower_shadow < body * 0.3 and last['close'] < last['open']:
        patterns.append('Shooting Star')
    
    # Doji
    if body < (last['high'] - last['low']) * 0.1:
        patterns.append('Doji')
    
    # Bullish Engulfing
    if (prev['close'] < prev['open'] and 
        last['close'] > last['open'] and
        last['open'] < prev['close'] and 
        last['close'] > prev['open']):
        patterns.append('Bullish Engulfing')
    
    # Bearish Engulfing
    if (prev['close'] > prev['open'] and 
        last['close'] < last['open'] and
        last['open'] > prev['close'] and 
        last['close'] < prev['open']):
        patterns.append('Bearish Engulfing')
    
    # Morning Star (3 candles)
    if prev2 is not None:
        if (prev2['close'] < prev2['open'] and
            abs(prev['close'] - prev['open']) < body * 0.3 and
            last['close'] > last['open'] and
            last['close'] > (prev2['open'] + prev2['close']) / 2):
            patterns.append('Morning Star')
    
    # Pin Bar
    total_range = last['high'] - last['low']
    if total_range > 0:
        if lower_shadow > total_range * 0.6 and body < total_range * 0.3:
            patterns.append('Bullish Pin Bar')
        elif upper_shadow > total_range * 0.6 and body < total_range * 0.3:
            patterns.append('Bearish Pin Bar')
    
    return patterns

def detect_chart_patterns(df):
    patterns = []
    
    if len(df) < 20:
        return patterns
    
    recent = df.tail(20)
    highs = recent['high'].values
    lows = recent['low'].values
    closes = recent['close'].values
    
    # Double Top
    peaks = []
    for i in range(1, len(highs) - 1):
        if highs[i] > highs[i-1] and highs[i] > highs[i+1]:
            peaks.append((i, highs[i]))
    
    if len(peaks) >= 2:
        last_two_peaks = peaks[-2:]
        if abs(last_two_peaks[0][1] - last_two_peaks[1][1]) < last_two_peaks[0][1] * 0.02:
            patterns.append('Double Top')
    
    # Double Bottom
    troughs = []
    for i in range(1, len(lows) - 1):
        if lows[i] < lows[i-1] and lows[i] < lows[i+1]:
            troughs.append((i, lows[i]))
    
    if len(troughs) >= 2:
        last_two_troughs = troughs[-2:]
        if abs(last_two_troughs[0][1] - last_two_troughs[1][1]) < last_two_troughs[0][1] * 0.02:
            patterns.append('Double Bottom')
    
    # Ascending Triangle
    if len(peaks) >= 2 and len(troughs) >= 2:
        if troughs[-1][1] > troughs[-2][1] and abs(peaks[-1][1] - peaks[-2][1]) < peaks[-1][1] * 0.01:
            patterns.append('Ascending Triangle')
    
    # Head and Shoulders (simplified)
    if len(peaks) >= 3:
        if peaks[-2][1] > peaks[-1][1] and peaks[-2][1] > peaks[-3][1]:
            if abs(peaks[-1][1] - peaks[-3][1]) < peaks[-1][1] * 0.02:
                patterns.append('Head and Shoulders')
    
    return patterns

File: server/services/patternDetector.js
class PatternDetector {
  detectCandlestickPatterns(candles) {
    const patterns = [];
    const len = candles.length;

    if (len < 3) return patterns;

    const current = candles[len - 1];
    const prev = candles[len - 2];
    const prev2 = candles[len - 3];

    // Hammer
    if (this.isHammer(current)) {
      patterns.push({ name: 'Hammer', bullish: true, strength: 0.7 });
    }

    // Shooting Star
    if (this.isShootingStar(current)) {
      patterns.push({ name: 'Shooting Star', bullish: false, strength: 0.7 });
    }

    // Doji
    if (this.isDoji(current)) {
      patterns.push({ name: 'Doji', bullish: null, strength: 0.5 });
    }

    // Bullish Engulfing
    if (this.isBullishEngulfing(prev, current)) {
      patterns.push({ name: 'Bullish Engulfing', bullish: true, strength: 0.85 });
    }

    // Bearish Engulfing
    if (this.isBearishEngulfing(prev, current)) {
      patterns.push({ name: 'Bearish Engulfing', bullish: false, strength: 0.85 });
    }

    // Morning Star
    if (this.isMorningStar(prev2, prev, current)) {
      patterns.push({ name: 'Morning Star', bullish: true, strength: 0.9 });
    }

    // Evening Star
    if (this.isEveningStar(prev2, prev, current)) {
      patterns.push({ name: 'Evening Star', bullish: false, strength: 0.9 });
    }

    // Three White Soldiers
    if (len >= 3 && this.isThreeWhiteSoldiers(candles.slice(-3))) {
      patterns.push({ name: 'Three White Soldiers', bullish: true, strength: 0.9 });
    }

    // Pin Bar
    if (this.isPinBar(current)) {
      patterns.push({ name: 'Pin Bar', bullish: current.close > current.open, strength: 0.75 });
    }

    return patterns;
  }

  isHammer(candle) {
    const body = Math.abs(candle.close - candle.open);
    const lowerShadow = Math.min(candle.open, candle.close) - candle.low;
    const upperShadow = candle.high - Math.max(candle.open, candle.close);
    
    return lowerShadow > body * 2 && upperShadow < body * 0.3;
  }

  isShootingStar(candle) {
    const body = Math.abs(candle.close - candle.open);
    const upperShadow = candle.high - Math.max(candle.open, candle.close);
    const lowerShadow = Math.min(candle.open, candle.close) - candle.low;
    
    return upperShadow > body * 2 && lowerShadow < body * 0.3;
  }

  isDoji(candle) {
    const body = Math.abs(candle.close - candle.open);
    const range = candle.high - candle.low;
    return body < range * 0.1;
  }

  isBullishEngulfing(prev, current) {
    return prev.close < prev.open && 
           current.close > current.open &&
           current.open < prev.close &&
           current.close > prev.open;
  }

  isBearishEngulfing(prev, current) {
    return prev.close > prev.open && 
           current.close < current.open &&
           current.open > prev.close &&
           current.close < prev.open;
  }

  isMorningStar(first, second, third) {
    return first.close < first.open &&
           Math.abs(second.close - second.open) < (first.high - first.low) * 0.3 &&
           third.close > third.open &&
           third.close > (first.open + first.close) / 2;
  }

  isEveningStar(first, second, third) {
    return first.close > first.open &&
           Math.abs(second.close - second.open) < (first.high - first.low) * 0.3 &&
           third.close < third.open &&
           third.close < (first.open + first.close) / 2;
  }

  isThreeWhiteSoldiers(candles) {
    return candles.every(c => c.close > c.open) &&
           candles[1].close > candles[0].close &&
           candles[2].close > candles[1].close;
  }

  isPinBar(candle) {
    const body = Math.abs(candle.close - candle.open);
    const totalRange = candle.high - candle.low;
    const nose = candle.close > candle.open ? 
                 candle.high - candle.close : 
                 candle.high - candle.open;
    
    return nose > totalRange * 0.6 && body < totalRange * 0.3;
  }

  detectChartPatterns(candles) {
    const patterns = [];

    // Double Top/Bottom
    const doublePattern = this.detectDoubleTopBottom(candles);
    if (doublePattern) patterns.push(doublePattern);

    // Head and Shoulders
    const hsPattern = this.detectHeadAndShoulders(candles);
    if (hsPattern) patterns.push(hsPattern);

    // Triangle
    const trianglePattern = this.detectTriangle(candles);
    if (trianglePattern) patterns.push(trianglePattern);

    return patterns;
  }

  detectDoubleTopBottom(candles) {
    if (candles.length < 20) return null;

    const highs = candles.map(c => c.high);
    const lows = candles.map(c => c.low);
    
    // Simple double top detection
    const recentHighs = highs.slice(-20);
    const maxHigh = Math.max(...recentHighs);
    const highIndices = recentHighs.map((h, i) => h > maxHigh * 0.98 ? i : -1).filter(i => i !== -1);
    
    if (highIndices.length >= 2 && highIndices[highIndices.length - 1] - highIndices[0] > 5) {
      return { name: 'Double Top', bullish: false, strength: 0.8 };
    }

    return null;
  }

  detectHeadAndShoulders(candles) {
    // Simplified H&S detection
    if (candles.length < 30) return null;
    
    const recent = candles.slice(-30);
    const highs = recent.map(c => c.high);
    const maxIdx = highs.indexOf(Math.max(...highs));
    
    if (maxIdx > 10 && maxIdx < 20) {
      const leftShoulder = Math.max(...highs.slice(0, maxIdx - 5));
      const rightShoulder = Math.max(...highs.slice(maxIdx + 5));
      const head = highs[maxIdx];
      
      if (head > leftShoulder * 1.05 && head > rightShoulder * 1.05) {
        return { name: 'Head and Shoulders', bullish: false, strength: 0.85 };
      }
    }
    
    return null;
  }

  detectTriangle(candles) {
    if (candles.length < 20) return null;
    
    const recent = candles.slice(-20);
    const highs = recent.map(c => c.high);
    const lows = recent.map(c => c.low);
    
    const highRange = Math.max(...highs) - Math.min(...highs);
    const lowRange = Math.max(...lows) - Math.min(...lows);
    
    if (highRange < lowRange * 0.5) {
      return { name: 'Ascending Triangle', bullish: true, strength: 0.75 };
    }
    
    return null;
  }
}

module.exports = PatternDetector;
